src/Filesystem/Provider/FileStreamAccess.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+namespace ILIAS\Filesystem\Provider;
+
+use ILIAS\Filesystem\Exception\FileAlreadyExistsException;
+use ILIAS\Filesystem\Exception\FileNotFoundException;
+use ILIAS\Filesystem\Exception\IOException;
+use Psr\Http\Message\StreamInterface;
```

Comment:
I backtrack regarding RAII. I think you understand it correctly, though. But since we have less guarantees on the point where objects are distructed in PHP (wrt C++, where RAII comes from), the pattern is not applicable.

Replacing the stream interface is not too essential to me, though.